src/Env.hs

```diff
@@ -0,0 +1,320 @@
+checkAndGet k m pos = case M.lookup k m of
+  Nothing -> Left ("Error in " ++ (sourcePosPretty pos) ++ " : " ++ k ++ " Doesn't exist in the context \n")
+  Just v ->  Right v
+
+lookUpK :: VarEnv -> Arg -> K
+lookUpK e (AT  (TTypeVar t))  = (Ktype ((typeVarMap e) M.! t)) --(Ktype (Type {typeName = "type", typePos = typeVarPos t }))
+lookUpK e (AT  (TConstr t)) =
+        if (nameCons t) `elem` (declaredNames e)
+        then lookUpK e (AVar (VarConst (nameCons t)))
+        else (Ktype (Type { typeName = "type", typePos = constructorInvokerPos t }))
+lookUpK e (AVar v) = (KT ((varMap e) M.! v))
+
+getTypesOfArgs :: VarEnv -> [Arg] -> [K]
+getTypesOfArgs e args = map (lookUpK e) args
+
+updateEnv :: VarEnv -> (Y, K) -> VarEnv
+updateEnv e (TypeVarY y, Ktype t) = e { typeVarMap = M.insert y t $ typeVarMap e }
+updateEnv e (VarY y, KT t)        = e { varMap = M.insert y t $ varMap e }
+updateEnv e err                   = e { errors = (errors e) ++ "Problem in update: " ++ (show err) ++ "\n" }
+
+addName :: String -> VarEnv -> VarEnv
+addName a e = if (a `elem` (names e))
+              then e {errors = (errors e) ++ ("Name already exsist in the context \n")}
+              else e {names = a : (names e)}
+
+addDeclaredName :: String -> VarEnv -> VarEnv
+addDeclaredName a e = if (a `elem` (declaredNames e))
+                      then e { errors = (errors e) ++ ("Name already exsist in the context \n") }
+                      else e { declaredNames = a : (declaredNames e) }
+
+--------------------------------------- Env Data Types ---------------------------------------
+
+-- | Environment for the dsll semantic checker. As the 'check' function executes, it
+-- accumulate information such as symbol tables in the environment.
+
+-- | list of elements that might appear in the global context
+
+data Ttype = Ttype { yt :: Y,
+                     kt :: K }
+             deriving (Show, Eq, Typeable)
+
+data TypeConstructor = TypeConstructor { nametc :: String,
+                                         klstc  :: [K],
+                                         typtc  :: Type }
+                       deriving (Show, Eq, Typeable)
+
+data VarConstructor = VarConstructor { namevc :: String,
+                                       ylsvc  :: [Y],
+                                       klsvc  :: [K],
+                                       tlsvc  :: [T],
+                                       tvc    :: T }
+                      deriving (Show, Eq, Typeable)
+
+data Operation = Operation { nameop :: String,
+                             ylsop  :: [Y],
+                             klsop  :: [K],
+                             tlsop  :: [T],
+                             top  :: T}
+                 deriving (Show, Eq, Typeable)
+
+data PredicateEnv = Pred1 Predicate1
+                  | Pred2 Predicate2
+                  deriving (Show, Eq, Typeable)
+
+data Predicate1 = Predicate1 { namepred1 :: String,
```

Comment:
confusing to use the same name twice -- name the constructor `Pred1` (and same for `Pred2`)